src/frontend/src/limiter.rs

```diff
@@ -0,0 +1,291 @@
+        debug!(
+            "processing write bytes: {}, current in-flight write bytes: {}",
+            processing_write_bytes,
+            in_flight_write_bytes.load(Ordering::Relaxed)
+        );
+        Self {
+            in_flight_write_bytes,
+            processing_write_bytes,
+        }
+    }
+}
+
+impl Drop for InFlightWriteBytesCounter {
+    // When the request is finished, the in-flight write bytes should be decreased.
+    fn drop(&mut self) {
+        self.in_flight_write_bytes
+            .fetch_sub(self.processing_write_bytes, Ordering::Relaxed);
+    }
+}
+
+impl Limiter {
+    pub fn new(max_in_flight_write_bytes: u64) -> Self {
+        Self {
+            max_in_flight_write_bytes,
+            in_flight_write_bytes: Arc::new(AtomicU64::new(0)),
+        }
+    }
+
+    pub fn limit_request(&self, request: &Request) -> Option<InFlightWriteBytesCounter> {
+        let size = match request {
+            Request::Inserts(requests) => self.insert_requests_data_size(requests),
+            Request::RowInserts(requests) => self.rows_insert_requests_data_size(requests),
+            _ => 0,
+        };
+        self.limit_in_flight_write_bytes(size as u64)
+    }
+
+    pub fn limit_row_inserts(
+        &self,
+        requests: &RowInsertRequests,
+    ) -> Option<InFlightWriteBytesCounter> {
+        let size = self.rows_insert_requests_data_size(requests);
+        self.limit_in_flight_write_bytes(size as u64)
+    }
+
+    /// Returns None if the in-flight write bytes exceed the maximum limit.
+    /// Otherwise, returns Some(InFlightWriteBytesCounter) and the in-flight write bytes will be increased.
+    pub fn limit_in_flight_write_bytes(&self, bytes: u64) -> Option<InFlightWriteBytesCounter> {
+        let result = self.in_flight_write_bytes.fetch_update(
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+            |current| {
+                if current + bytes > self.max_in_flight_write_bytes {
+                    warn!(
+                        "in-flight write bytes exceed the maximum limit {}, request with {} bytes will be limited",
+                        self.max_in_flight_write_bytes,
+                        bytes
+                    );
+                    return None;
+                }
+                Some(current + bytes)
+            },
+        );
+
+        match result {
+            // Update the in-flight write bytes successfully.
+            Ok(_) => Some(InFlightWriteBytesCounter::new(
+                self.in_flight_write_bytes.clone(),
+                bytes,
+            )),
+            // It means the in-flight write bytes exceed the maximum limit.
+            Err(_) => None,
+        }
+    }
+
+    /// Returns the current in-flight write bytes.
+    #[allow(dead_code)]
+    pub fn in_flight_write_bytes(&self) -> u64 {
+        self.in_flight_write_bytes.load(Ordering::Relaxed)
+    }
+
+    fn insert_requests_data_size(&self, request: &InsertRequests) -> usize {
```

Comment:
I'm wondering if it's possible to read the payload size from protocol layer instead of calculating them ourselves. An extra pass to walk the entire data might be costly.